airflow/providers/cncf/kubernetes/utils/pod_manager.py

```diff
@@ -91,6 +99,61 @@ def get_container_termination_message(pod: V1Pod, container_name: str):
         return container_status.state.terminated.message if container_status else None
 
 
+class PodLogsConsumer:
+    """
+    PodLogsConsumer is responsible for pulling pod logs from a stream with checking a container status before
+    reading data.
+    This class is a workaround for the issue https://github.com/apache/airflow/issues/23497
+    """
+
+    def __init__(
+        self,
+        response: HTTPResponse,
+        pod: V1Pod,
+        pod_manager: PodManager,
+        container_name: str,
+        timeout: int = 120,
+    ):
+        self.response = response
+        self.pod = pod
+        self.pod_manager = pod_manager
+        self.container_name = container_name
+        self.timeout = timeout
+
+    def __iter__(self) -> Generator[bytes, None, None]:
+        messages: list[bytes] = []
+        if self.logs_available():
+            for chunk in self.response.stream(amt=None, decode_content=True):
+                if b"\n" in chunk:
+                    chunks = chunk.split(b"\n")
+                    yield b"".join(messages) + chunks[0] + b"\n"
+                    for x in chunks[1:-1]:
+                        yield x + b"\n"
+                    if chunks[-1]:
+                        messages = [chunks[-1]]
+                    else:
+                        messages = []
+                else:
+                    messages.append(chunk)
+                if not self.logs_available():
```

Comment:
this would seem to result in a _lot_ of calls to the kube API.  it may be wise to limit such cheking to every couple minutes?  wdyt? @jedcunningham ?

Reply:
This whole approach seems a little backwards to me, if I'm understanding things correctly. _Sometimes_ we can't continue fetching logs after 2 minutes (or whatever), so we _always_ stop trying then?

Would setting a timeout on the request itself guard us from getting stuck forever? Maybe `_request_timeout` on `read_namespaced_pod_log`? If not, I feel like we should find a way to bail once we know we've hit it instead of just not attempting it in the first place.

Reply:
> Sometimes we can't continue fetching logs after 2 minutes (or whatever), so we always stop trying then?

I think what's happening Jed is we always stop 2 minutse _after container completes_.

Reply:
> _request_timeout

I think that what's different here is author only applies timeout in the "container done" scenario.  But `_request_timeout` would be across the whole request, which if following, could be a long time.

Reply:
> I think what's happening Jed is we always stop 2 minutse after container completes.
> I think that what's different here is author only applies timeout in the "container done" scenario. But _request_timeout would be across the whole request, which if following, could be a long time.

@dstandish , you got it right.

Limiting these calls by 2 minutes is a good idea. I'll update PR along with other fixes.